File: num_meth/sem6/lab3/implicit_euler.f90
module kinds
  implicit none
  integer, parameter :: dp=kind(0.d0)  !  double precision
end module kinds
!*******************************************************************************

module nonlinear
  use kinds
  implicit none
  private 
  public :: euler 
contains

  !*******************************************************************************
  subroutine euler()
    integer :: k=0
    integer, parameter :: n=2
    real(dp), parameter :: T=1
    real(dp) :: x(n), y_k(n), f_vec(n)
    real(dp) :: t_k=0.00000001
    real(dp) :: tau_k

    y_k(1) = 0
    y_k(2) = -0.412

    do while(t_k < T)
       f_vec = f(y_k, t_k)
       tau_k = step(f_vec)
       y_k = y_k + tau_k * f_vec
       t_k = t_k + tau_k
       print*,
       write(*,*) "K: ", k
       print*, "Solution: ["
       write(*,*) "U1 = ",y_k(1)
       write(*,*) "U2 = ",y_k(2)
       print*, "]"
       write(*,*) "T: ", t_k
       print*, "---------------------------"
       k = k + 1
    end do
    
  end subroutine euler

  !******************************************************
  function step(f)
    integer :: i
    integer, parameter :: n=2
    real(dp), parameter :: e_dop = 0.00001
    real(dp), parameter :: tau_max = 0.01
    real(dp), intent(in) :: f(n)
    real(dp) :: tau(n)
    real(dp) :: step

    do i=1,n
       tau(i) = e_dop / abs(f(i)) + e_dop / tau_max
    end do
    step = minval(tau)
  end function step

  !******************************************************
  function f(u,t)
    !  input function
    implicit none
    integer, parameter :: n=2
    integer, parameter :: Om=40
    real(dp) :: a
    real(dp), intent(in) :: u(n)
    real(dp), intent(in) :: t
    real(dp) :: f(n)

    a = 2.5
    a = a + Om/40

    ! f(1) = -u(1)*u(2) + sin(t) / t
    ! f(2) = -(u(2) * u(2)) + a*t/(1+t*t)

    f(1) = -2 * u(1) + 4 * u(2)
    f(2) = -u(1) + 3 * u(2)
 
  end function f

end module nonlinear
!*******************************************************************************

program main
  use kinds
  use nonlinear
  implicit none 
  call euler()
end program main

File: num_meth/sem6/lab3/implicit_euler.py
import numpy as np
import copy

e_dop = 0.00001
u_0 = np.array([0, -0.412])
u_0_test = np.array([3, 0])
Om = 40
a = 2.5 + Om/40
tau_max = 0.01
T = 1

NUMBER_OF_EQUATIONS = 2

def step(f):
    tau = []
    for i in range(NUMBER_OF_EQUATIONS):
        tau.append(e_dop / np.abs(f[i]) + e_dop / tau_max)
    return min(tau)

def f(u, t):
    f = np.zeros([NUMBER_OF_EQUATIONS])
    f[0] = -u[0]*u[1] + np.sin(t) / t
    f[1] = -(u[1] ** 2) + a*t/(1+t**2)
    return f

def f_test(u, t):
    f = np.zeros([NUMBER_OF_EQUATIONS])
    f[0] = -2*u[0] + 4*u[1]
    f[1] = -u[0] + 3*u[1]
    return f

def euler():
    t_k = 0.00000001
    y_k = u_0
    k = 0
    while t_k < T:
        f_vec = f(y_k, t_k)
        tau_k = step(f_vec)
        y_k = y_k + tau_k * f_vec
        t_k += tau_k        
        print('K: [{2}] | Y: {0} | T: {1}'.format(y_k, t_k, k))
        k += 1


if __name__ == '__main__':
    euler()
